### PR TITLE
Support auth with both JWTs and API Keys

### PR DESCRIPTION
**Pull request by flesler**
Connects-to: #1 
Change-type: major

This new project `resin-auth` comes to replace `resin-token`. It adds support for api keys using the same interface and changes somewhat the way the string key is preserved, in here both `JWT` and `APIKey` instances encapsulate the string, each time it is changed, new instances are created. `resin-auth`, unlike `resin-token` doesn't expose `getData()`. We decided that the SDK will always query the API for the user data, even if a JWT is used.
---- Autogenerated Waffleboard Connection: Connects to #1

### CODE REVIEW

File: lib/auth.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+interface ResinAuthOptions {
+	dataDirectory?: string;
+	tokenKey?: string;
+}
+
+export class ResinAuth {
+	private readonly storage: ResinSettingsStorage;
+	private readonly tokenKey: string;
+	private token?: Token;
```
[the remark this answers is not in the document]
Comment (flesler) [reply lib/auth.ts]: It is true that it is, in some way a cache. I still prefer this name, but willing to change if majority thinks it's better.

As for getting out of sync, it shouldn, how would that happen?

File: lib/auth.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+			throw new errors.ResinMalformedToken(key);
+		}
+		if (token.isExpired()) {
+			throw new errors.ResinExpiredToken(token.getKey());
+		}
+		return token;
+	};
+
+	private getToken = (): Promise<Token> => {
+		if (this.token) {
```
[the remark this answers is not in the document]
Comment (flesler) [reply lib/auth.ts]: It makes sense but, if it's stored we know it's valid and as for expired, I'm pretty sure we defined with @pimterry that we'd let the API validate that, that has been the behavior so far AFAIK

File: README.md [hunk 3]
```diff
@@ -48,6 +48,8 @@ const auth = require('resin-auth')({
 })
 ```
 
+ERROR, Cannot find module.
```
[the remark this answers is not in the document]
Comment (flesler) [reply README.md]: Wasn't sure what was up with it but assumed it would be fixed once the rest of the stuff is ok. Don't worry I won't publish it like this. Do you know why isn't it finding the module? Does it fetch from NPM or sth like that?

File: lib/api-key.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+ */
+
+import { Token, TokenType } from './token';
+
+export class APIKey implements Token {
+	private readonly key: string;
+
+	constructor(key: string) {
+		this.key = key;
+	}
```
[the remark this answers is not in the document]
Comment (flesler) [reply lib/api-key.ts]: The question is, should I? hehe
I don't quite like that alternative.

File: lib/api-key.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	 * @summary Get the type of token
+	 * @function
+	 * @public
+	 *
+	 * @returns {TokenType} the type of token
+	 *
+	 * @example
+	 * console.log(token.getType())
+	 */
+	public getType = (): TokenType => TokenType.APIKey;
```
[the remark this answers is not in the document]
Comment (flesler) [reply lib/api-key.ts]: I thought interfaces were only about functions but seems like this could be specified as a property in the `Token` interface. Will try the change

File: tests/01-jwt.spec.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
+import { chai } from 'mochainon';
```
[the remark this answers is not in the document]
Comment (flesler) [reply tests/01-jwt.spec.ts]: Yes, I want the jwt and api-key tests to run first. So if sth breaks within them, the error will show in the right place and not in the auth tests

File: tests/03-auth.spec.ts [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+import jwtFixtures from './fixtures/jwts';
+
+const { expect } = chai;
+const IS_BROWSER = typeof window !== 'undefined';
+
+let dataDirectory;
+if (!IS_BROWSER) {
+	// tslint:disable-next-line no-var-requires
+	const settings = require('resin-settings-client');
+	dataDirectory = settings.get('dataDirectory');
```
[the remark this answers is not in the document]
Comment (flesler) [reply tests/03-auth.spec.ts]: TBH this part remains exactly as it is in [resin-token](https://github.com/resin-io-modules/resin-token/blob/master/tests/token.spec.ts#L11) so I think it's acceptable? What do you think?

File: tests/03-auth.spec.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+const auth = new ResinAuth({ dataDirectory });
+
+describe('ResinAuth', () => {
+	before(() =>
```
[the remark this answers is not in the document]
Comment (flesler) [reply tests/03-auth.spec.ts]: No, not really.

File: tests/03-auth.spec.ts [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+		});
+	});
+
+	// These only check one basic variation of each, the other specs go into more detail
+
+	describe('.setKey()', () => {
+		describe('JWT', () => {
+			before(() => auth.setKey(jwtFixtures.empty));
+
+			after(() => auth.removeKey());
```
[the remark this answers is not in the document]
Comment (flesler) [reply tests/03-auth.spec.ts]: If in the future we tests several JWTs here, I think we should have an inner `describe` and it will have its own set of `before` and `after`.

File: tests/03-auth.spec.ts [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+			});
+
+			describe('.isValid()', () => {
+				it('should return true for a valid JWT', () => {
+					expect(auth.isValid()).to.eventually.equal(true);
+				});
+			});
+
+			describe('.getAge()', () => {
+				it('should return 0 when there is no `iat`', () => {
```
[the remark this answers is not in the document]
Comment (flesler) [reply tests/03-auth.spec.ts]: I'll check `resin-token`, the idea is to remain compatible

[the remark this answers is not in the document]
Comment (flesler) [reply tests/03-auth.spec.ts]: @pimterry correct. It seems [resin-token](https://github.com/resin-io-modules/resin-token/blob/master/lib/token.ts#L318) is returning (implicitely) `undefined` when there is no `iat`. Are you ok with an `undefined` there to handle it on the SDK or do you prefer a `throw`?

File: tests/03-auth.spec.ts [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+			describe('.needs2FA()', () => {
+				it('should return false when there is no `twoFactorRequired`', () => {
+					expect(auth.needs2FA()).to.eventually.equal(false);
+				});
+			});
+		});
+
+		describe('APIKey', () => {
+			before(() => auth.setKey(apiKeyFixtures.apiKey));
```
[the remark this answers is not in the document]
Comment (flesler) [reply tests/03-auth.spec.ts]: It's not necessary. As commented above in this file, this module tests a single case just to make sure the whole promise wrapping works fine. All the scenarios for each are tested on the other specs. I was a bit hesitant on whether to just test the whole thing from the outside (`ResinAuth`) or what. Thoughts?

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (pimterry) [reply README.md]: Nope, it's purely local. It means the jsdoc docs config in this module is broken. I don't really understand exactly how that works, but you should be able to crib from the SDK, the CLI, or most of our other modules, which all do similar things.

File: lib/api-key.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (pimterry) [reply lib/api-key.ts]: Up to you! :smiley: It's functionally exactly the same thing, just more terse.

File: tests/01-jwt.spec.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (pimterry) [reply tests/01-jwt.spec.ts]: I'm not sure what this means.

File: tests/03-auth.spec.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (pimterry) [reply tests/03-auth.spec.ts]: I'm sure there was somewhere where we changed it, to solve exactly this issue... Even if we haven't, I'm reasonably sure that's the effect of this, and it will be kind of annoying that every time you run these tests they log you out of the CLI, for example.

File: tests/03-auth.spec.ts [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+	const settings = require('resin-settings-client');
+	dataDirectory = settings.get('dataDirectory');
+}
+
+const auth = new ResinAuth({ dataDirectory });
+
+describe('ResinAuth', () => {
+	before(() =>
+		// Ensure a clean state before starting
+		auth.removeKey()
```
Comment (pimterry) [inline tests/03-auth.spec.ts]: This is a promise - we should return it, so the test doesn't start until the token is definitely removed

File: tests/03-auth.spec.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (pimterry) [reply tests/03-auth.spec.ts]: If you make this a beforeEach though, you guarantee every test starts with a clean slate, and you make it way easier to ensure these don't start depending on one another later on. And since this should be an unnoticeably quick operation, there's no real downside.

As a bonus, if you make this one a beforeEach, you can remove the `after` that I commented on below completely, because you'll have guaranteed that every test gets an empty token before it starts anyway.

File: tests/03-auth.spec.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (pimterry) [reply tests/03-auth.spec.ts]: Ooooh, interesting. I'd forgotten to think about compatibility here, good catch. Let's go with undefined: compatibility is nice, and it is arguably semantically the right choice (that's what's actually happening: the age is undefined).

File: tests/03-auth.spec.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (pimterry) [reply tests/03-auth.spec.ts]: At the moment, this is making assumptions about what all the subsequent tests will do, that it doesn't need to. If you just add those extra 4 characters, you guarantee that even if one of the below tests goes crazy and mutates something, or if you add another one later on that intentionally does so without thinking about this, then everything will still be fine. It's a free safeguard.

> I was a bit hesitant on whether to just test the whole thing from the outside (ResinAuth) or what. Thoughts?

I'm not sure exactly what this means. I do think this test structure, and the specific tests you're covering here are sensible, in combination with the other unit tests.

File: lib/api-key.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (flesler) [reply lib/api-key.ts]: I rather have them separated

File: tests/01-jwt.spec.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (flesler) [reply tests/01-jwt.spec.ts]: The `auth` spec relies on the underlying `jwt` and `api-key` modules. If we ever break something in those modules, it's very likely that tests on `auth` spec will also break. In that case, I want to make sure they break first on the "low-level" modules' spec. It's not that relevant because we are not bailing on the first error but still.

TL;DR: In my opinion the ordering of tests here matter and I want those 2 spec modules to run before the `auth` one

File: tests/03-auth.spec.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (flesler) [reply tests/03-auth.spec.ts]: I agree. Instead of changing the `dataDirectory` I changed the `tokenKey` so they are stored in the same directory with a different key

File: tests/03-auth.spec.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (flesler) [reply tests/03-auth.spec.ts]: Ok done.

File: tests/03-auth.spec.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (flesler) [reply tests/03-auth.spec.ts]: Done

File: tests/03-auth.spec.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (flesler) [reply tests/03-auth.spec.ts]: Ok changed to return `undefined` when there's no `iat`

File: tests/03-auth.spec.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (flesler) [reply tests/03-auth.spec.ts]: Applied your suggestions

File: tests/03-auth.spec.ts (hunk 12 above)
Comment (flesler) [reply tests/03-auth.spec.ts]: It is, there are no braces

Comment (pimterry) [reply tests/03-auth.spec.ts]: Weird, I could've sworn I checked and saw them... I'm going mad, clearly.

File: tests/01-jwt.spec.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (pimterry) [reply tests/01-jwt.spec.ts]: I don't really agree with this - test order really shouldn't ever matter, and it's not what we do everywhere else - but I don't care very much, and I'm totally happy to ignore it :smiley:

File: tests/03-auth.spec.ts [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+	);
+
+	describe('.hasKey()/.removeKey()', () => {
+		it('should return false when no key was stored', () => {
+			expect(auth.hasKey()).to.eventually.equal(false);
+		});
+
+		it('should return true when any key was successfully stored', () => {
+			auth.setKey(jwtFixtures.empty).then(() => {
```
Comment (thgreasi) [inline tests/03-auth.spec.ts]: I think that we need to return that promise.

File: tests/03-auth.spec.ts [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+				expect(auth.hasKey()).to.eventually.equal(true);
+			});
+		});
+
+		it('should return false again when the key is removed', () => {
+			auth
```
Comment (thgreasi) [inline tests/03-auth.spec.ts]: I think that we need to return that promise as well.

File: tests/03-auth.spec.ts [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+				.setKey(jwtFixtures.empty)
+				.then(() => {
+					auth.removeKey();
```
Comment (thgreasi) [inline tests/03-auth.spec.ts]: I think that we need to return that promise as well.

File: tests/03-auth.spec.ts (hunk 13 above)
Comment (thgreasi) [inline tests/03-auth.spec.ts]: Do we need a return here?

File: tests/03-auth.spec.ts (hunk 14 above)
Comment (thgreasi) [inline tests/03-auth.spec.ts]: Do we need a return here?

File: tests/03-auth.spec.ts (hunk 15 above)
Comment (thgreasi) [inline tests/03-auth.spec.ts]: Do we need a return here?

File: lib/api-key.ts [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+	 * @returns {boolean} is valid
+	 *
+	 * @example
+	 * console.log(token.isValid());
+	 */
+	public isValid = (): boolean => true;
+
+	/**
+	 * @member getAge
+	 * @summary get the token age
```
Comment (thgreasi) [inline lib/api-key.ts]: NIT: Just as a heads-up, some summaries start with a capital while others with lowercase.

File: tests/03-auth.spec.ts (hunk 13 above)
Comment (flesler) [reply tests/03-auth.spec.ts]: Nice catch man, I mean the tests pass in both cases and with the same tests run. But it should be needed, updated

File: tests/03-auth.spec.ts (hunk 14 above)
Comment (flesler) [reply tests/03-auth.spec.ts]: Ditoo

File: tests/03-auth.spec.ts (hunk 15 above)
Comment (flesler) [reply tests/03-auth.spec.ts]: Ditto

File: lib/api-key.ts (hunk 16 above)
Comment (flesler) [reply lib/api-key.ts]: Fixed
